pyfar/_codec.py

```diff
@@ -0,0 +1,296 @@
+        zip-archive. This zip-path is stored as a reference together with a
+        type-hint as a pair into the JSON-form
+            [str, str] e.g. ['$ndarray', '/my_obj/_signal']
+
+Class-Level
+===========
+
+Actually pyfar_obj.__dict__ could be passed to `_encode` or `_decode`
+respectively. Some objects however need special treatment at the highest level
+of encoding and decoding. Therefore `PyfarClass._encode` and
+`PyfarClass._decode` must be implemented in each class. E.g.:
+
+class MyClass:
+
+    def _encode(self):
+        return self.copy().__dict__
+
+    @classmethod
+    def _decode(cls, obj_dict):
+        obj = cls()
+        obj.__dict__.update(obj_dict)
+        return obj
+
+Data Inspection
+===============
+
+Once data is written to disk you can rename the file-extension to .zip, open
+and inspect the archive. The JSON-file in the archive must reflect the
+data structure, e.g. like this:
+
+JSON
+----
+{
+    "_n": 42,
+    "_comment": "My String",
+    "_subobj": {
+        "signal": [
+            "ndarray",
+            "my_obj/_subobj/signal"
+        ],
+        "_m": 49
+    }
+    "_list": [
+        1,
+        [
+            "dtype",
+            "int32"
+        ],
+    ]
+}
+----
+
+"""
+
+import io
+import sys
+import numpy as np
+
+
+def _decode(obj, zipfile):
+    """
+    This function is exclusively used by `io.read` and enables recursive
+    decoding for objects of varying depth.
+
+    Parameters
+    ----------
+    obj : PyFar-object.
+
+    zipfile: zipfile-object.
+        The zipfile object is looped in the recursive structure
+        e.g. to decode ndarrays when they occur.
+    """
+    if isinstance(obj, dict):
+        for key in obj.keys():
+            _inner_decode(obj, key, zipfile)
+    elif isinstance(obj, list):
+        for i in range(0, len(obj)):
+            _inner_decode(obj, i, zipfile)
+
+    return obj
+
+
+def _inner_decode(obj, key, zipfile):
+    """
+    This function is exclusively used by `io._decode`
+
+    Parameters
+    ----------
+    obj : PyFar-object.
+
+    key :  str or int
+        The key provided by the dict or list over which currently is being
+        iterated.
+    """
+    if not _is_type_hint(obj[key]):
+        _decode(obj[key], zipfile)
+    elif _is_pyfar_type(obj[key][0][1:]):
+        ParfyType = _str_to_type(obj[key][0][1:])
+        obj[key] = ParfyType._decode(obj[key][1])
+        _decode(obj[key].__dict__, zipfile)
+    elif obj[key][0][1:] == 'dtype':
+        obj[key] = getattr(np, obj[key][1])
+    elif obj[key][0][1:] == 'ndarray':
+        obj[key] = _decode_ndarray(obj[key][1], zipfile)
+    else:
+        _decode_numpy_scalar(obj, key)
+        pass
+
+
+def _decode_numpy_scalar(obj, key):
+    """ This function is exclusively used by `io._inner_decode` and
+    decodes numpy scalars e.g. of type `numpy.int32`.
+    """
+    try:
+        numpy_scalar = getattr(np, obj[key][0][1:])
+    except AttributeError:
+        pass
```

Comment:
This is because in `_inner_decode` I do not ask if it's a numpy-type, rather I leave this to `getattr(np, obj[key][0][1:])`. If it's not a numpy-type the exception is caught and the value is left as is (usually a bultin). see EAFP.